### PR TITLE
Validate an image release

### LINKED ISSUE #172: Assume --all if --update-image is provided with no --service
> **jml** just did a deploy w/ flux earlier today. I don't know if it's helpful feedback, but I found the command `fluxctl --url=http://localhost:8001/api/v1/proxy/namespaces/fluxy/services/fluxy release --update-image=quay.io/weaveworks/ui-server:master-139c12d --all` much harder to remember than the equivalent `wcloud deploy quay.io/weaveworks/ui-server:master-139c12d` command. Had to consult the docs.

Perhaps we can default to assuming --all if --update-image is provided with no explicit --service?

### LINKED ISSUE #232: releasing an invalid image results in 'Complete ... Nothing to do'
I accidentally specified just the image tag instead of image tag + name, and here's what fluxctl said:
```
$ ./fluxctl -u http://frontend.dev.weave.works/api/flux -t <token> release --service=scope/control --update-image=2002-goroutine-leak-in-control-0610936
Submitting release job...
Release job submitted, ID 133d47bd-2153-6a4f-8d60-9995687defa3
Status: Complete.

Here's what happened:
 1) Submitted job.
 2) Calculating release actions.
 3) Release 2002-goroutine-leak-in-control-0610936 to scope/control
 4) All selected services are running the requested images. Nothing to do.
Took 4.728496s
```

I would have expected it to complain that the image doesn't exist, or perhaps that the image name is invalid.

### LINKED ISSUE #257: Unusual experience rolling out authfe
Apologies in advance that this isn't a well-defined bug. I just wanted to capture an unusual experience rolling out authfe w/ fluxctl in case it helps.

Doing the daily prod roll out, a copy/paste error meant that I accidentally typed `fluxctl release --all --update-image quay.io/weaveworks/authfe:quay.io/weaveworks/authfe` and hit enter.

I hit Ctrl-C as quickly as I could. I got some sort of output, which unfortunately I no longer have access to. 

I then ran the correct command:

```
$ fluxctl release --all --update-image quay.io/weaveworks/authfe:master-8de2e90
```

Which emitted 3 lines of output (again, I don't have it), and hanged.

I went to our `compare-images` page showed production running the revision `quay.io/weaveworks/authfe`.

During this time, I could still access admin pages which were behind the authfe.

Due to time pressure (I can't leave prod broken for long; others were waiting for deploy to clear), I attempted to resolve the problem by running `wcloud deploy quay.io/weaveworks/authfe:master-8de2e90`, which worked.

### PR DESCRIPTION
**Pull request by philwinder**
By:
- Checking that the requested image is valid (i.e. has a tag)
- Checking that the image exists in the repository.

Fixes #172
Fixes #232 
Fixes #257

### CODE REVIEW

File: instance/instance_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	testImageExists(t, i, exampleImage, true)
+	testImageExists(t, i, "owner/repo", false)
+	testImageExists(t, i, "owner:tag", false)
+	testImageExists(t, i, "", false)
+}
+
+func testImageExists(t *testing.T, i Instance, image string, expected bool) {
+	b, err := i.imageExists(flux.ParseImageID(image))
+	if err != nil {
+		t.Fatal(err.Error())
```
Comment (paulbellamy) [inline instance/instance_test.go]: Add the images into these error messages, so it is easier to tell what failed.

File: service.go [hunk 2]
```diff
@@ -212,11 +212,19 @@ func (s ServiceSpec) AsID() (ServiceID, error) {
...
+		return ImageSpec(s), nil
 	}
-	return ImageSpec(ParseImageID(s))
+	id := ParseImageID(s)
+	_, name, tag := id.Components()
+	if name == "" {
+		return "", errors.Wrap(ErrInvalidImageID, "blank image name")
+	}
+	if tag == "" {
+		return "", errors.Wrap(ErrInvalidImageID, "blank tag (if you want latest, explicitly state the tag :latest")
```
Comment (paulbellamy) [inline service.go]: missed closing parens.

File: instance/instance.go [hunk 3]
```diff
@@ -125,11 +125,37 @@ func (h *Instance) GetRepository(repo string) ([]flux.ImageDescription, error) {
...
 		m[id.Repository()] = []flux.ImageDescription{flux.ImageDescription{ID: id}}
 	}
 	return m, nil
 }
 
+// Checks whether the given image exists in the repository.
+// Return true if exist, false otherwise
+func (h *Instance) imageExists(image flux.ImageID) (exists bool, err error) {
+	// Get a list of images
+	images, err := h.registry.GetRepository(image.Repository())
```
Comment (paulbellamy) [inline instance/instance.go]: Bit inefficient to fetch the entire repo and all tags just to check one.

File: instance/instance_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	descriptions []flux.ImageDescription
+	err          error
+}
+
+func NewMockRegistry(descriptions []flux.ImageDescription, err error) registry.Client {
+	return &mockRegistry{
+		descriptions: descriptions,
+		err:          err,
+	}
+}
```
Comment (paulbellamy) [inline instance/instance_test.go]: I suspect we'll want this in other packages' tests, so might be worth putting in `/registry/mock.go`

File: instance/instance.go (hunk 3 above)
Comment (philwinder) [inline instance/instance.go]: Will need major refactor in order to do this. Will do this in separate PR.

**Comment by philwinder** [comment]
Unable to test due to repository structure. Will add tests for this after refactor.

**Pull request merged**